syncing/src/main/kotlin/maru/syncing/pipeline/DownloadCompleteBlockRangeTask.kt

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Consensys Software Inc.
+ *
+ * This file is dual-licensed under either the MIT license or Apache License 2.0.
+ * See the LICENSE-MIT and LICENSE-APACHE files in the repository root for details.
+ *
+ * SPDX-License-Identifier: MIT OR Apache-2.0
+ */
+package maru.syncing.pipeline
+
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.TimeoutException
+import maru.core.SealedBeaconBlock
+import maru.p2p.PeerLookup
+import org.apache.logging.log4j.LogManager
+import tech.pegasys.teku.networking.p2p.reputation.ReputationAdjustment
+
+class DownloadCompleteBlockRangeTask(
+  private val peerLookup: PeerLookup,
+) {
+  private val log: org.apache.logging.log4j.Logger = LogManager.getLogger(this::javaClass)
+
+  fun getCompleteBlockRange(targetRange: SyncTargetRange): CompletableFuture<List<SealedBeaconBlock>> =
+    CompletableFuture<List<SealedBeaconBlock>>.supplyAsync {
+      var startBlockNumber = targetRange.startBlock
+      var count = targetRange.endBlock - targetRange.startBlock + 1uL
+      var remaining = count
+      val downloadedBlocks = mutableListOf<SealedBeaconBlock>()
+      var retries = 0
+      do {
+        val peer = peerLookup.getPeers().random() // TODO: filter peers? Do we want to use least busy peer?
+        try {
+          peer
+            .sendBeaconBlocksByRange(startBlockNumber, remaining)
```

Comment:
How is `remaining` supposed to work in practice? What if the number of blocks doesn't fit into the recipient's memory, but the sender won't mind sending massive number of blocks over the network?

Or to put the question differently, aren't we supposed to throttle the count with some upper bound on the client side?

Reply:
There is a max defined for the number of blocks that you can request.
I guess we do select a default number of blocks that we request at a time, which cannot be greater than  max.
We should implement throttling in another PR.